module/zfs/dmu_recv.c

```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+
+		buf = kmem_alloc(lsize, KM_SLEEP);
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(os, obj, offset, lsize, buf,
+		    DMU_READ_NO_PREFETCH);
+		kmem_free(buf, lsize);
+		if (err != ECKSUM)
+			goto cleanup; /* no corruption found */
+
+		err = dmu_buf_hold_noread(os, obj, offset, FTAG, &dbp);
+		if (err != 0) {
+			err = SET_ERROR(EBUSY);
+			goto cleanup;
+		}
+		blkid =
+		    dbuf_whichblock(DB_DNODE((dmu_buf_impl_t *)dbp), 0, offset);
+		/* Get the block pointer for the corrupted block */
+		bp = dmu_buf_get_blkptr(dbp);
+		dmu_buf_rele(dbp, FTAG);
+
+		break;
+	}
+	case DRR_SPILL:
+	{
+		dnode_t *dn;
+		dmu_buf_t *dbp_spill;
+		obj = drr.drr_u.drr_spill.drr_object;
+		lsize = drr.drr_u.drr_spill.drr_length;
+		csize = drr.drr_u.drr_spill.drr_compressed_size;
+
+		err = dmu_bonus_hold(os, obj, FTAG, &dbp);
+		if (err != 0)
+			goto cleanup;
+
+		/* Try to read the spill block to see if it needs healing */
+		err = dmu_spill_hold_existing(dbp, FTAG, &dbp_spill);
+		if (err != ECKSUM) {
+			dmu_buf_rele(dbp, FTAG);
+			if (err == 0)
+				dmu_buf_rele(dbp_spill, FTAG);
+			goto cleanup; /* no corruption found */
+		}
+		/* Get the block pointer to the corrupted spill block */
+		dn = dmu_buf_dnode_enter(dbp);
+		bp = DN_SPILL_BLKPTR(dn->dn_phys);
+		dmu_buf_dnode_exit(dbp);
+		dmu_buf_rele(dbp, FTAG);
+		break;
+	}
+	default:
+		ASSERT0(1);
+		err = SET_ERROR(EINVAL);
+		goto cleanup;
+	}
+
+	/* Get the good data from the recv record */
+	abd = abd_get_from_buf(rrd->arc_buf->b_data,
+	    arc_buf_size(rrd->arc_buf));
+	abd_take_ownership_of_buf(abd, B_FALSE);
+
+	if (arc_get_compression(rrd->arc_buf) != BP_GET_COMPRESS(bp)) {
+		/*
+		 * The compression in the stream doesn't match what we had
+		 * on disk; we need to re-compress the buf into the
+		 * compression type that was written to disk previously.
+		 */
+		abd_t *decompressed_abd;
+		buf = kmem_alloc(lsize, KM_SLEEP);
+		/* Decompress the stream data */
+		err = zio_decompress_data(arc_get_compression(rrd->arc_buf),
+		    abd, buf, csize, lsize);
+		if (err != 0) {
+			kmem_free(buf, lsize);
+			abd_release_ownership_of_buf(abd);
+			abd_put(abd);
+			goto cleanup;
+		}
+
+		decompressed_abd = abd_get_from_buf(buf, lsize);
+		abd_take_ownership_of_buf(decompressed_abd, B_FALSE);
+		/* Recompress the stream data */
+		VERIFY3U(zio_compress_data(BP_GET_COMPRESS(bp),
+		    decompressed_abd, buf, lsize), ==, csize);
+		size = csize;
+
+		/* Swap in the newly compressed data into the abd */
+		abd_release_ownership_of_buf(decompressed_abd);
+		abd_put(decompressed_abd);
+		abd_release_ownership_of_buf(abd);
+		abd_put(abd);
+		abd = abd_get_from_buf(buf, csize);
+		abd_take_ownership_of_buf(abd, B_FALSE);
+
+		kmem_free(buf, lsize);
+	} else {
+		size = arc_buf_size(rrd->arc_buf);
+	}
+
+	/* Correct the corruption in place */
+	err = zio_wait(zio_rewrite(NULL, os->os_spa, 0, bp, abd, size, NULL,
+	    NULL, ZIO_PRIORITY_SYNC_WRITE, flags, NULL));
```

Comment:
Given the current functionality, we would expect there to be a LOT more records for non-corrupt blocks, than for corrupt blocks.  So I'm even more concerned about the synchronous (and not eligible for predictive prefetch) `dmu_read()`'s to determine if we want to correct this block.

We'd typically (e.g. with >1 vdev) get better performance by simply always asynchronously `zio_rewrite()`ing every record (i.e. without checking if its actually corrupt).  Since you'd have Nrecords async writes vs the current PR's Nrecords sync reads.  Plus the code would be a lot simpler.

I wonder if we should wait for the extensions (to send only the corrupt blocks) are ready, or at least design `recv -c` to work best in that mode.  E.g. by assuming that nearly all records will be for corrupt blocks?